### PR TITLE
Added Functions to extend chart functionality

### PR DESCRIPTION
**Pull request by pete-pjb**
Added Functions to extend chart functionality as follows:

/**
 * Set the index of the x-axis start point in the data array
 * @param chart             pointer to a chart object
 * @param ser 				pointer to a data series on 'chart'
 * @param id    			the index of the x point in the data array
 */
void lv_chart_set_x_start_point(lv_obj_t * chart, lv_chart_series_t * ser, uint16_t id);

/**
 * Set an external array of data points to use for the chart
 * NOTE: It is the users responsibility to make sure the point_cnt matches the external array size.
 * @param chart             pointer to a chart object
 * @param ser 				pointer to a data series on 'chart'
 * @param array				external array of points for chart
 */
void lv_chart_set_ext_array(lv_obj_t * chart, lv_chart_series_t * ser, lv_coord_t array[], uint16_t point_cnt );

/**
 * Set an individual point value in the chart series directly based on index
 * @param chart             pointer to a chart object
 * @param ser 				pointer to a data series on 'chart'
 * @param value				value to assign to array point
 * @param id				the index of the x point in the array
 */
void lv_chart_set_point_id(lv_obj_t * chart, lv_chart_series_t * ser, lv_coord_t value, uint16_t id);

/**
 * get the current index of the x-axis start point in the data array
 * @param chart             pointer to a chart object
 * @param ser 				pointer to a data series on 'chart'
 * @return 					the index of the current x start point in the data array
 */
uint16_t lv_chart_get_x_start_point(lv_obj_t * chart, lv_chart_series_t * ser);

/**
 * Get an individual point value in the chart series directly based on index
 * @param chart             pointer to a chart object
 * @param ser 				pointer to a data series on 'chart'
 * @param id				the index of the x point in the array
 * @return					value of array point at index id
 */
lv_coord_t lv_chart_get_point_id(lv_obj_t * chart, lv_chart_series_t * ser, uint16_t id);

### CODE REVIEW

File: src/lv_widgets/lv_chart.c [hunk 1]
```diff
@@ -303,43 +303,44 @@ void lv_chart_set_point_count(lv_obj_t * chart, uint16_t point_cnt)
...
+				LV_ASSERT_MEM(ser->points);
+				if(ser->points == NULL) return;
+				/*Initialize the new points*/
+				if(point_cnt > point_cnt_old) {
+					for(i = point_cnt_old - 1; i < point_cnt; i++) {
+						ser->points[i] = def;
+					}
+				}
+			}
+    	}
```
Comment (kisvegabor) [inline src/lv_widgets/lv_chart.c:343]: Can you add an `LV_LOG_WARN` in the else branch to indicate that nothing will happen?

**Review by kisvegabor — COMMENTED** [review]
Looks good thank you!
I've just spotted a memory leak the chart and fixed it here: eacc9d8c

In `LV_SIGNAL_CLEANUP` `ser->points` should be freed only if the data is not external.
Could you update this PR from the latest master add this check?

File: src/lv_widgets/lv_chart.c (hunk 1 above)
Comment (pete-pjb) [reply src/lv_widgets/lv_chart.c:343]: Hi @kisvegabor ,

Sorry about missing the signal handler, it's just my current unfamiliarity with the internals of LVGL.  I have made a change to that now.

Before I update the pull request can I ask where you want the LV_LOG_WARN adding please, it isn't clear to me?

Kind Regards,

Pete

Comment (kisvegabor) [reply src/lv_widgets/lv_chart.c:343]: > Sorry about missing the signal handler, it's just my current unfamiliarity with the internals of LVGL. I have made a change to that now.

No problem, this part wasn't even there 2 hours ago :slightly_smiling_face: 

> Before I update the pull request can I ask where you want the LV_LOG_WARN adding please, it isn't clear to me?

I mean something like this in `lv_chart_set_point_count`:
```c
if( !ser->ext_buf_assigned ) {
 ...
} else {
  LV_LOG_WARN("Can't set the point coint if external array is used");
}
```

Comment (pete-pjb) [reply src/lv_widgets/lv_chart.c:343]: I see, but currently it is still setting the point count even with the external buffer.   I am likely to use this to chart various densities of points from the same buffer. Looking through the code it should not cause any problem I can see.  Is that okay?  If not I can just move the point count setting into the conditional and add the else with the log.  I could just do another call to assign buffer function for my purposes if you would rather do things that way....

 What do you think?

Comment (kisvegabor) [reply src/lv_widgets/lv_chart.c:343]: Upps, sorry. The bottom of the function was hidden in GitHub's diff and didn't notice ` ext->point_cnt = point_cnt;` there.

All good from my side.

**Review by kisvegabor — APPROVED** [review]

**Review by embeddedt — APPROVED** [review]

**Pull request merged**